bmeg_app/components/func_repurpose_drug.py

```diff
@@ -0,0 +1,177 @@
+        disease[row[0]]=row[3]
+        if row[0] not in data: 
+            data[row[0]] = { row[1] :  row[2] } 
+        else: 
+            data[row[0]][row[1]] = row[2]  
+    drugDF = pd.DataFrame(data).transpose()
+    drugDF = drugDF.sort_index().sort_index(axis=1) #sort rows by cell line name, sort cols by drug name
+    return drugDF, disease
+
+def compare_drugs(est,drugDF,disease,fig_yaxisLabel, selected_disease):
+    '''
+    ex. compare_drugs('PACLITAXEL',drugDF,disease,'Breast Cancer')
+    '''
+    import plotly.graph_objects as go
+    import pandas as pd
+    # exclude non-breast cancer derived cell lines
+    new_col=[]
+    for ind in drugDF.index:
+        new_col.append(disease[ind])
+    drugDF['disease']=new_col
+    drugDF=drugDF[drugDF['disease']==selected_disease]
+    # set established drug to first col
+    new_col=drugDF.pop(est)
+    drugDF.insert(0, est, new_col)
+    drugDF.pop('disease')
+    # Melt df (cols: cellline, drug, value)
+    melt_drugDF= drugDF
+    melt_drugDF['Cell Line']=melt_drugDF.index
+    melt_drugDF= melt_drugDF.melt(id_vars=['Cell Line'])
+    melt_drugDF.columns=['Cell Line', 'Drug', 'Drug Response']
+    # Generate violin plots
+    fig = go.Figure()
+    drugs = list(set(melt_drugDF['Drug']))
+    assert est in drugs
+    drugs.remove(est)
+    drugs.insert(0,est) # place established drug at index0
+
+    for d in drugs:
+        fig.add_trace(go.Violin(x=melt_drugDF['Drug'][melt_drugDF['Drug'] == d],
+                                y=melt_drugDF['Drug Response'][melt_drugDF['Drug'] == d],
+                                name=d,box_visible=True,meanline_visible=True))
+    fig.update_layout(margin={'t':10, 'b':10},height=400,yaxis=dict(title=fig_yaxisLabel))
+    fig.update_xaxes(tickangle=45)
+    fig.update_layout(showlegend=False)
+    return melt_drugDF, fig
+
+
+def mappings(selected_project):
+    if 'CCLE' == selected_project:
+        q = G.query().V('Project:CCLE').out("cases").as_("p").out("samples").out("aliquots").out("drug_response").as_("dr").out("compounds").as_("c")
+        drugRes = q.aggregate(gripql.term("drugs", '_data.synonym')).execute()
+        drug_opts = [a['key'].upper() for a in drugRes[0]['drugs']['buckets']]
+        drug_opts = {}
+        for a in drugRes[0]['drugs']['buckets']:
+            drug_opts[a['key'].upper()]= a['key']
+        return drug_opts
+    #elif other project...load project specific properties
+def mappings_drugResp(selected_project):
+    if 'CCLE' == selected_project:
+        return {
+        'AAC':'$dr._data.aac',
+        'IC50':'$dr._data.ic50',
+        'EC50': '$dr._data.ec50'
+         }
+    #elif other project...load project specific properties
+
+def drugDetails(drugs_list):
+    '''
+    create table of drugs and their taxon.
+    '''
+    a=[]
+    b=[]
+    c=[]
+    d=[]
+    e=[]
+    f=[]
+    g=[]
+    for DRUG in drugs_list:
+        q=G.query().V().hasLabel('Compound').has(gripql.eq('_data.synonym', DRUG))
```

Comment:
This is a very slow way to query the data. Every time this is called, the database scan through all compounds looking for a matching Synonym, which is not indexed. If the GID was stored, and used to access these records, the database would be able to go directly to the record, without scanning the entire table. It would also allow you to pull multiple GIDs at the same time with a single query.

Reply:
Passed list of GID into query to utilize indexing